### PR TITLE
Improve channel list and model conversion performance

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🔗 Issue Links

- [IOS-912](https://linear.app/stream/issue/IOS-912/channel-list-synchronize-takes-a-lot-of-time-when-multiple-members-per)
- https://linear.app/stream/issue/IOS-1397/v5-reduce-swiftui-sdk-size-by-investigating-usage-of-chatchannel-and

### 🎯 Goal

Reduce the CPU cost of loading and updating the channel list, especially for channels with many members and long message histories. The work targets the database-to-model conversion path and repeated main-thread work observed while profiling the channel list.

### 📝 Summary

- Convert `ChatChannel` and `ChatMessage` to final, immutable reference types to avoid repeated value copies during conversions.
- Prewarm a cache for channel read lookups to avoid repeated scans.
- Skip building and discarding latest messages beyond the truncation window.
- Avoid rebuilding entity models twice per background database change.
- Skip decoding extra data when it is an empty JSON object.
- Avoid a redundant queue hop when reading `chatClientConfig`.

### 🛠 Implementation

The main change is making `ChatChannel`/`ChatMessage` final classes, which removes a class of redundant copying in the DB observers and conversion helpers. The remaining commits each remove specific redundant work on the conversion / background-observer path (double model rebuilds, unnecessary message materialization past the truncation window, empty extra-data decoding, and a queue hop).

### 🎨 Showcase

N/A — no user-facing UI changes.

### 🧪 Manual Testing Notes

Load a channel list where channels have many members and long histories, and confirm channel list synchronize/scroll no longer hangs the main thread. Behavior should be unchanged.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change should be manually QAed
- [x] Changelog is updated with client-facing changes
- [ ] Changelog is updated with new localization keys
- [x] New code is covered by unit tests
- [ ] Documentation has been updated in the `docs-content` repo

## Summary

* **New Features**
  * Added persisted, stable identifiers for channel reads with pre-warming and automatic backfilling to prevent duplicate entries.
* **Bug Fixes**
  * Improved membership state updates for channel membership events.
  * Fixed latest-message selection so truncation and oldest limits apply consistently.
* **Performance**
  * Expanded cached-item reuse across channel, user, message, poll, member, and related observers for faster updates.
  * Faster decoding for empty custom JSON payloads.
* **Breaking/Behavioral Changes**
  * `ChatChannel` and `ChatMessage` are now reference types; `changing(...)` behavior was updated.
* **Tests**
  * Added/expanded coverage for read identifiers, cache reuse, observer behavior, and decoder edge cases.

### CODE REVIEW

File: Sources/StreamChat/APIClient/Endpoints/Payloads/IdentifiablePayload.swift [hunk 1]
```diff
@@ -143,7 +143,13 @@ extension ChannelPayload: IdentifiablePayloadProxy {
...
         messages.fillIds(cache: &cache)
         pinnedMessages.fillIds(cache: &cache)
-        channelReads.fillIds(cache: &cache)
+        // A channel read's id is composed of the channel cid and the user id, so it can only be
+        // resolved here where the parent cid is known.
+        channelReads.forEach { read in
+            read.fillIds(cache: &cache)
+            let readId = ChannelReadDTO.createId(cid: channel.cid, userId: read.user.id)
+            cache[ChannelReadDTO.className, default: []].insert(readId)
+        }
```
Comment (laevandus) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/IdentifiablePayload.swift]: Would be nice to have the fillIds wrapper to keep the callsite similar to others.

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift [hunk 2]
```diff
@@ -2075,7 +2075,8 @@ private extension ChatChannelController {
                 fetchRequest: ChannelDTO.fetchRequest(for: cid),
                 itemCreator: {
                     try $0.asModel() as ChatChannel
-                }
+                },
+                itemReuseKeyPaths: (\ChatChannel.cid.rawValue, \ChannelDTO.cid)
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:2079]: We need to do additional testing with channel controller because this relies on CoreData detecting changes correctly when submodels also change (eg willSave we add here and there).

Never tried but there is an option of changing model reuse implementation and not use list changes and compare final ChatChannels instead inside the observer. More CPU work, but might be more correct. But outside of scope of this PR.

Let's add these to state layer as well for consistency.

File: Sources/StreamChat/Controllers/ChannelController/LivestreamChatHandler.swift [hunk 3]
```diff
@@ -360,18 +360,16 @@ final class LivestreamChatHandler: LivestreamChatHandling, DataStoreProvider, @u
...
             members.insert(memberAddedEvent.member)
             let memberCount = channel?.memberCount ?? 0
 
-            var membership: ChatChannelMember?
-            if memberAddedEvent.member.id == currentUserId {
-                membership = memberAddedEvent.member
-            }
+            let membership: ChatChannelMember?? = memberAddedEvent.member.id == currentUserId
+                ? .some(memberAddedEvent.member)
+                : nil
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/ChannelController/LivestreamChatHandler.swift:372]: Double optional looks a bit odd, for readability it would be nice to avoid it. WDYT?

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 4]
```diff
@@ -586,16 +586,18 @@ extension ChatChannel {
...
                 .prefix(clientConfig.localCaching.chatChannel.latestMessagesLimit)
+                .prefix(while: { $0.createdAt.bridgeDate >= lowerBound })
                 .compactMap { try? $0.relationshipAsModel(depth: depth) }
-            if let oldest = dto.oldestMessageAt?.bridgeDate {
-                messages = messages.filter { $0.createdAt >= oldest }
-            }
-            if let truncated = dto.truncatedAt?.bridgeDate {
-                messages = messages.filter { $0.createdAt >= truncated }
-            }
-            return messages
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/ChannelDTO.swift:598]: Makes sense, good change to reduce model conversions.

File: Sources/StreamChat/Database/DTOs/ChannelReadDTO.swift [hunk 5]
```diff
@@ -53,18 +61,35 @@ class ChannelReadDTO: NSManagedObject {
...
+        if let existing = load(id: readId, context: context) {
             return existing
         }
 
+        // Fallback for reads persisted before the `id` attribute existed: look them up by the
+        // compound predicate and backfill the identifier so subsequent lookups can use the id.
+        if let legacy = load(cid: cid, userId: userId, context: context) {
+            legacy.id = readId
+            return legacy
+        }
```
Comment (laevandus) [inline Sources/StreamChat/Database/DTOs/ChannelReadDTO.swift]: Just wondering if this could be skipped if CoreData model has id as non-optional and then does the DatabaseContainer recreatePersistentStore get automatically invoked or not. It would flush away old data then.

File: Sources/StreamChat/Database/DatabaseSession.swift [hunk 6]
```diff
@@ -7,12 +7,14 @@ import CoreData
...
+    /// wrapping the read in `performAndWait` only adds a redundant queue-confinement check. This getter
+    /// is called very frequently (once per model created while decoding a channel/message/member), so
+    /// avoiding that overhead matters.
     var chatClientConfig: ChatClientConfig? {
-        nonisolated(unsafe) var config: ChatClientConfig?
-        performAndWait {
-            config = userInfo[Self.chatClientConfigKey] as? ChatClientConfig
-        }
-        return config
+        userInfo[Self.chatClientConfigKey] as? ChatClientConfig
```
Comment (laevandus) [inline Sources/StreamChat/Database/DatabaseSession.swift]: This was added in #3566. I wonder if things have changed since then and it is really safe now.

File: Sources/StreamChat/Controllers/ChannelController/LivestreamChatHandler.swift (hunk 3 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/LivestreamChatHandler.swift:372]: This one is actually needed because we need to distinguish between:
- There is a value
- There is no value
- The value must be reset to nil

Found it weird as well initially, but it is actually a clean solution for this particular case

File: Sources/StreamChat/Database/DTOs/ChannelReadDTO.swift (hunk 5 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/ChannelReadDTO.swift]: Yeah I guess this might not be needed, since we don't have migrations. I will review this.

File: Sources/StreamChat/Database/DatabaseSession.swift (hunk 6 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DatabaseSession.swift]: Good catch, will check

File: Sources/StreamChat/APIClient/Endpoints/Payloads/IdentifiablePayload.swift (hunk 1 above)
Comment (nuno-vieira) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/IdentifiablePayload.swift]: Done 👍

File: Sources/StreamChat/Database/DTOs/ChannelReadDTO.swift (hunk 5 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DTOs/ChannelReadDTO.swift]: Removed 👍

File: Sources/StreamChat/Database/DatabaseSession.swift (hunk 6 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Database/DatabaseSession.swift]: Reverted this since it was risky and the impact is not huge

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 2 above)
Comment (martinmitrevski) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:2079]: good point about the state layer, especially if we want to use it under the hood here as well

**Review by martinmitrevski — APPROVED** [review]
these changes look good to me - but let's test both UIKit and SwiftUI, cc @testableapple. Will review, test and comment now on the SwiftUI one.

File: Sources/StreamChat/Controllers/ChannelController/ChannelController.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController.swift:2079]: Added already 👍

**Review by laevandus — APPROVED** [review]
